src/Forms/TreeMultiselectField.php

```diff
@@ -119,48 +120,66 @@ public function getSchemaStateDefaults()
-            if (is_array($sourceObject)) {
-                $values = is_array($this->value) ? $this->value : preg_split('/ *, */', trim($this->value));
+        $value = $this->Value();
 
-                foreach ($values as $value) {
-                    $item = new stdClass;
-                    $item->ID = $value;
-                    $item->Title = $sourceObject[$value];
-                    $items->push($item);
-                }
-                return $items;
+        // If unchanged, load from record
+        if ($value === 'unchanged') {
+            // Verify a form exists
+            $form = $this->getForm();
+            if (!$form) {
+                return ArrayList::create();
             }
 
-            // Otherwise, look data up from the linked relation
-            if (is_string($this->value)) {
-                $ids = explode(',', $this->value);
-                foreach ($ids as $id) {
-                    if (!is_numeric($id)) {
-                        continue;
-                    }
-                    $item = DataObject::get_by_id($sourceObject, $id);
-                    if ($item) {
-                        $items->push($item);
-                    }
-                }
-                return $items;
+            // Verify this form has an attached record with the necessary relation
+            $fieldName = $this->getName();
+            $record = $form->getRecord();
+            if ($record instanceof DataObject && $record->hasMethod($fieldName)) {
+                return $record->$fieldName();
             }
+
+            // No relation on parent record found
+            return ArrayList::create();
         }
 
-        if ($this->form) {
-            $fieldName = $this->name;
-            $record = $this->form->getRecord();
-            if (is_object($record) && $record->hasMethod($fieldName)) {
-                return $record->$fieldName();
+        // Value is a list
+        if ($value instanceof SS_List) {
+            return $value;
+        }
+
+        // Parse ids from value string / array
+        $ids = [];
+        if (is_string($value)) {
+            $ids = preg_split("#\s*,\s*#", trim($value));
+        } elseif (is_array($value)) {
+            $ids = array_values($value);
+        }
+
+        // No value
+        if (empty($ids)) {
+            return ArrayList::create();
+        }
+
+        // Query source records by value field
+        return DataObject::get($this->getSourceObject())
+            ->filter($this->getKeyField(), $ids);
+    }
+
+    public function setValue($value, $source = null)
+    {
+        // If loading from a dataobject, get items by relation
+        if ($source instanceof DataObject) {
+            $name = $this->getName();
+            if ($source->hasMethod($name)) {
+                $value = $source->$name();
             }
         }
 
-        return $items;
+        return parent::setValue($value);
+    }
+
+    public function dataValue()
```

Comment:
docblock here would be good, even if it just linked to https://github.com/silverstripe/silverstripe-framework/blob/4.6/src/Forms/FormField.php#L478